Font Library: sanitize font collection data

matiasbenedetto opened the pull request:
## What?
Font Library: sanitize font collection data.
- Removes all the potential unwanted properties.
- Sanitizes all the values.
- Adds unit test case to ensure that data returned by  `get_data` is always sanitized.

## Why?
To return just the data that's safe to return.

## How?
- Adding a data schema in the WP_Font_Collection class.
- Sanitizing the data using the sanitizing util from WP_Font_Utils

## Testing Instructions
Run this PHP snippet featuring a font collection with risky data:

```php
function register_collections() {
    $ubuntu = array(
        'font_family_settings' => array (
            'name'       => 'Ubuntu<script>alert("xss")</script>',
            'fontFamily' => 'Ubuntu, sans-serif<script>alert("xss")</script>',
            'slug'       => 'ubuntu<script>alert("xss")</script>',
        ),
        'categories' => array(
            'sans-serif',
        ),
    );
    
    $verdana = array(
        'font_family_settings' => array (
            'name'       => 'Verdana<script>alert("xss")</script>',
            'fontFamily' => 'Verdana, sans-serif<script>alert("xss")</script>',
            'slug'       => 'verdana<script>alert("xss")</script>',
        ),
        'categories' => array(
            'sans-serif',
        ),
    );
    
    $font_families = array ( $ubuntu, $verdana );
    
    $categories = array(
        array (
            'name' => 'Sans Serif<script>alert("xss")</script>',
            'slug' => 'sans-serif<script>alert("xss")</script>',
        ),
    );
    
    $collection_with_xss = array(
        'name'          => 'PHP Custom Collection',
        'description'   => __( 'Custom fonts collection' ),
        'font_families' => $font_families,
        'categories'    => $categories
    );
    
    wp_register_font_collection( 'collection-with-xss', $collection_with_xss );
}

add_action( 'rest_api_init', 'register_collections' );
```
Request that font collection using the API:
```
/wp-json/wp/v2/font-collections/collection-with-xss
```

the response should be:

```json
{
    "slug": "collection-with-xss",
    "name": "PHP Custom Collection",
    "description": "Custom fonts collection",
    "font_families": [
        {
            "font_family_settings": {
                "name": "Ubuntu",
                "fontFamily": "Ubuntu, sans-serif",
                "slug": "ubuntualertxss"
            },
            "categories": [
                "sans-serif"
            ]
        },
        {
            "font_family_settings": {
                "name": "Verdana",
                "fontFamily": "Verdana, sans-serif",
                "slug": "verdanaalertxss"
            },
            "categories": [
                "sans-serif"
            ]
        }
    ],
    "categories": [
        {
            "name": "Sans Serif",
            "slug": "sans-serifalertxss"
        }
    ],
    "_links": {
        "self": [
            {
                "href": "http://localhost/wp1/wp-json/wp/v2/font-collections/collection-with-xss"
            }
        ],
        "collection": [
            {
                "href": "http://localhost/wp1/wp-json/wp/v2/font-collections"
            }
        ]
    }
}
```

matiasbenedetto committed:
sanitize font collection data

matiasbenedetto committed:
add a test case to be sure that the data returned by get_data is always sanitized following the font collection schema

matiasbenedetto committed:
format php

matiasbenedetto committed:
lint

creativecoder commented on lib/compat/wordpress-6.5/fonts/class-wp-font-collection.php:
```diff
@@ -44,6 +44,54 @@ final class WP_Font_Collection {
+								'fontFamily'            => 'sanitize_text_field',
+								'fontStyle'             => 'sanitize_text_field',
+								'fontWeight'            => 'sanitize_text_field',
+								'src'                   => 'sanitize_url',
```
What about when src is an array?

matiasbenedetto replied:
Added a custom method to do that: https://github.com/WordPress/gutenberg/pull/58636/commits/216d2106cb56478eab2a3cfbe127bf28ba4d98e3

creativecoder commented on lib/compat/wordpress-6.5/fonts/class-wp-font-collection.php:
```diff
@@ -203,6 +251,7 @@ private function load_from_url( $url ) {
 		 * @return array|WP_Error Array of data if valid, otherwise a WP_Error instance.
 		 */
 		private function validate_data( $data ) {
```
Rename to `validate_and_sanitize` data for clarity?

matiasbenedetto replied:
renamed here: https://github.com/WordPress/gutenberg/pull/58636/commits/d07b787b1a931cf5a8e32789122069f38447850b

matiasbenedetto committed:
custom function to validate src properties

matiasbenedetto committed:
rename function

matiasbenedetto committed:
php format

creativecoder committed:
Updates how collection data is sanitized and validated

- Sanitize and validate as early as possible, so that appropriate notices are logged right away
- Use static method for sanitization schema so that we can use a closure rather than a public method for src sanitization
- Adds a check for WP_Font_Utils::sanitize_from_schema so that class callables like array( $this, 'sanitization_method'	) can be used
- Updates method name to indicate that sanitization is done first, as this might remove invalid data and affect the validation result

creativecoder commented on lib/compat/wordpress-6.5/fonts/class-wp-font-utils.php:
```diff
@@ -158,7 +158,7 @@ public static function sanitize_from_schema( $tree, $schema ) {
 
 				$is_value_array  = is_array( $value );
-				$is_schema_array = is_array( $schema[ $key ] );
+				$is_schema_array = is_array( $schema[ $key ] ) && ! is_callable( $schema[ $key ] );
```
Adding this check allows using a class instance method, like `array( $this, 'sanitization_method' )`.

creativecoder commented:
@matiasbenedetto I made an attempt to streamline this and pushed a commit to this branch:

- Sanitize and validate as early as possible, so that appropriate notices are logged right away.
- Use static method for sanitization schema rather than a class constant so that we can use a closure for src sanitization rather than needing to add a new public method to the class.
- Adds a check within WP_Font_Utils::sanitize_from_schema so that class callables like `array( $this, 'sanitization_method'	)` can also be used
- Updates method name to `sanitize_and_validate_data` to indicate that sanitization is done first, as it might remove invalid data and affect the validation result.

Feel free to revert if you don't agree with the changes. Otherwise I think this now looks ready to go.

creativecoder committed:
Fix potential nitpicks
